### ISSUE TITLE
Run Functional Tests: Test run stops with TaskCanceledException

### ISSUE THREAD

**Issue by SevenBitSignature** [Microsoft/vsts-tasks#2308]
I can't get the Run Functional Tests task to work with our real-life test suites.

- On-premise TFS 2015 Update 3
- VS Test Agent Deployment task deploys agent version 14.0.25420.1
- Run Functional Tests task: Test Selection is Test Plan
- Test Suite: Multiple selected

In total there are 400+ automated tests spanning three assemblies (5 tests + 19 tests + the vast majority of tests). The tests from the two smaller assemblies are successfully executed and their results are uploaded to TFS, then the test run stops.

Error in DTAExecutionHost.exe.log is:

E, 5568, 4, 2016/08/19, 17:16:03.450, MYTESTMACHINE\DTAExecutionHost, TestExecutionServiceCommandQueueListener.SendStopCommandToHandleException : An exception occured in queue listener. System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.VisualStudio.TestService.AgentExecutionHost.TestExecutionServiceCommandQueueListener.<DispatchAsync>d__7.MoveNext() Sending Stop command...

Please find attached an anonymized DTAExecutionHost.exe.log excerpt for some more context (I could provide the full log to a supporter but not in a public forum).

[test_log.txt](https://github.com/Microsoft/vsts-tasks/files/430021/test_log.txt)

**Comment by tanvi-soni** [comment]
Hi, 
You can send DTAExecutionHost.exe.log, build definition logs and a snapshot of your run functional test task to devops_tools@microsoft.com .

**Comment by SevenBitSignature** [comment]
What do you mean by "snapshot" of the task?

**Comment by SevenBitSignature** [comment]
I think I found the root cause. Nothing to do with number of assemblies nor number of tests. Some test cases in the test plan contained the same automated test ID (accidentally of course) -- once I corrected the associated test automation, the test run started running all tests.

All tests with duplicate IDs happen to be in the "main" assembly, so I guess that a misconfiguration as described prevents test execution of all tests in the affected assembly.

There is a very, very subtle hint in DTAExecutionHost.exe.log which brought me on the right track in the end:

I, 5568, 9, 2016/08/19, 17:14:23.726, MYTESTMACHINE\DTAExecutionHost, DefaultTestRunEvenHandler.OnTestRunMessage : VstestAdapter.Execution - RunMessage : An item with the same key has already been added.

Bottom line: I think this should actually not prevent from tests to run at all. At the very least, a clear error message must be logged.

**Comment by tanvi-soni** [comment]
Good to see that you are unblocked. We will look into it further at our end.
